spaceship/lib/spaceship/connect_api/api_client.rb

```diff
@@ -165,15 +165,16 @@ def with_asc_retry(tries = 5, &_block)
 
         if [500, 504].include?(status)
           msg = "Timeout received! Retrying after 3 seconds (remaining: #{tries})..."
-          raise TimeoutRetryError, msg
+          raise RetryError, msg
+        elsif status == 401
+          msg = "Token has expired, or has been revoked! Trying to refresh..."
+          puts(msg) if Spaceship::Globals.verbose?
+          @token.refresh!
+          raise RetryError, "Retrying after 3 seconds (remaining: #{tries})..."
         end
 
         return response
-      rescue UnauthorizedAccessError => error
-        # Catch unathorized access and re-raising
-        # There is no need to try again
-        raise error
-      rescue TimeoutRetryError => error
+      rescue RetryError => error
         tries -= 1
```

Comment:
I would rather add second retry for UnauthAccessError rather then removing 401 code from 
spaceship/spec/connect_api/api_client_spec.rb

Reply:
@DuMaM I've updated the code to add retries to the `UnauthorizedAccessError` block instead. Let me know what you think.

Reply:
Does it work now? I'm unable to validate this in any way.

Reply:
Hi @DuMaM, yes it still works as before. You can see logs of one of our builds, using the updated patch, below:

```
INFO [2021-12-07 10:39:01.79]: -----------------------------------------------------
INFO [2021-12-07 10:39:01.79]: Successfully uploaded package to App Store Connect. It might take a few minutes until it's visible online.
INFO [2021-12-07 10:39:01.79]: -----------------------------------------------------
INFO [2021-12-07 10:39:01.80]: Successfully uploaded the new binary to App Store Connect
INFO [2021-12-07 10:39:01.80]: If you want to skip waiting for the processing to be finished, use the `skip_waiting_for_build_processing` option
INFO [2021-12-07 10:39:01.80]: Note that if `skip_waiting_for_build_processing` is used but a `changelog` is supplied, this process will wait for the build to appear on AppStoreConnect, update the changelog and then skip the remaining of the processing steps.
INFO [2021-12-07 10:39:01.94]: Waiting for processing on... app_id: 927068353, app_version: 2.24.1, build_version: 297991, platform: IOS
WARN [2021-12-07 10:39:02.33]: Read more information on why this build isn't showing up yet - https://github.com/fastlane/fastlane/issues/14997
INFO [2021-12-07 10:39:02.33]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
                               // ... many identical lines ...
INFO [2021-12-07 10:45:38.50]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
INFO [2021-12-07 10:46:15.62]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
Authentication credentials are missing or invalid. - Provide a properly configured and signed bearer token, and make sure that it has not expired. Learn more about Generating Tokens for API Requests https://developer.apple.com/go/?id=api-generating-tokens
Token has expired or has been revoked! Trying to refresh...
INFO [2021-12-07 10:46:47.25]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
                               // ... many identical lines ...
INFO [2021-12-07 10:55:43.51]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
Authentication credentials are missing or invalid. - Provide a properly configured and signed bearer token, and make sure that it has not expired. Learn more about Generating Tokens for API Requests https://developer.apple.com/go/?id=api-generating-tokens
Token has expired or has been revoked! Trying to refresh...
INFO [2021-12-07 10:56:15.55]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
INFO [2021-12-07 10:56:47.13]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
INFO [2021-12-07 10:57:18.20]: Successfully finished processing the build 2.24.1 - 297991 for IOS
```

As you can see, the token either expired or, more likely, was revoked twice in this process and fastlane was able to recover from the error, refreshed the token and the job finished successfully. \o/

Reply:
Hi @DuMaM, yes it still works as before. You can see logs of one of our builds, using the updated patch, below:

```
INFO [2021-12-07 10:39:01.79]: -----------------------------------------------------
INFO [2021-12-07 10:39:01.79]: Successfully uploaded package to App Store Connect. It might take a few minutes until it's visible online.
INFO [2021-12-07 10:39:01.79]: -----------------------------------------------------
INFO [2021-12-07 10:39:01.80]: Successfully uploaded the new binary to App Store Connect
INFO [2021-12-07 10:39:01.80]: If you want to skip waiting for the processing to be finished, use the `skip_waiting_for_build_processing` option
INFO [2021-12-07 10:39:01.80]: Note that if `skip_waiting_for_build_processing` is used but a `changelog` is supplied, this process will wait for the build to appear on AppStoreConnect, update the changelog and then skip the remaining of the processing steps.
INFO [2021-12-07 10:39:01.94]: Waiting for processing on... app_id: xxxxxxxxx, app_version: 2.24.1, build_version: 297991, platform: IOS
WARN [2021-12-07 10:39:02.33]: Read more information on why this build isn't showing up yet - https://github.com/fastlane/fastlane/issues/14997
INFO [2021-12-07 10:39:02.33]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
                               // ... many identical lines ...
INFO [2021-12-07 10:45:38.50]: Waiting for the build to show up in the build list - this may take a few minutes (check your email for processing issues if this continues)
INFO [2021-12-07 10:46:15.62]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
Authentication credentials are missing or invalid. - Provide a properly configured and signed bearer token, and make sure that it has not expired. Learn more about Generating Tokens for API Requests https://developer.apple.com/go/?id=api-generating-tokens
Token has expired or has been revoked! Trying to refresh...
INFO [2021-12-07 10:46:47.25]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
                               // ... many identical lines ...
INFO [2021-12-07 10:55:43.51]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
Authentication credentials are missing or invalid. - Provide a properly configured and signed bearer token, and make sure that it has not expired. Learn more about Generating Tokens for API Requests https://developer.apple.com/go/?id=api-generating-tokens
Token has expired or has been revoked! Trying to refresh...
INFO [2021-12-07 10:56:15.55]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
INFO [2021-12-07 10:56:47.13]: Waiting for App Store Connect to finish processing the new build (2.24.1 - 297991) for IOS
INFO [2021-12-07 10:57:18.20]: Successfully finished processing the build 2.24.1 - 297991 for IOS
```

As you can see, the token either expired or, more likely, was revoked twice in this process and fastlane was able to recover from the error, refreshed the token and the job finished successfully. \o/

Reply:
Nice to hear :D maybe it solves my issues now ;)
I think you should fix CI build errors and ping code owners to review this.